bee-trace/bee-trace/src/subscriber/layer/log.rs

```diff
@@ -0,0 +1,273 @@
+    Stdout(StdoutLock<'a>),
+    /// Log to a file.
+    File(MutexGuard<'a, File>),
+}
+
+impl<'a> io::Write for LogOutput<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self {
+            Self::Stdout(lock) => lock.write(buf),
+            Self::File(lock) => lock.write(buf),
+        }
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        match self {
+            Self::Stdout(lock) => lock.flush(),
+            Self::File(lock) => lock.flush(),
+        }
+    }
+}
+
+/// Describes the target destination of a [`log`] event.
+///
+/// Locks obtained from these targets are used to create writers to the appropriate [`LogOutput`].
+enum LogDest {
+    /// Log to standard output.
+    Stdout,
+    /// Log to a file.
+    File(Mutex<File>),
+}
+
+/// Describes a target destination of a [`log`] event, combined with filters that only permit
+/// specific events to be logged to that target.
+struct LogTarget {
+    /// Target filters. Enables/disables [`Span`]s based on their target and level.
+    filter: Targets,
+    /// The output destination of the event, if it passes through the filter.
+    dest: LogDest,
+}
+
+/// [`MakeWriter`] implementation for the [`LogLayer`].
+///
+/// Constructs a writer for a specific [`LogTarget`].
+struct LogTargetMakeWriter {
+    stdout: Stdout,
+    target: LogTarget,
+}
+
+impl LogTargetMakeWriter {
+    fn new(target: LogTarget) -> Self {
+        Self {
+            stdout: io::stdout(),
+            target,
+        }
+    }
+
+    fn enabled<S>(&self, meta: &Metadata<'_>, ctx: &Context<'_, S>) -> bool
+    where
+        S: tracing::Subscriber + for<'a> LookupSpan<'a>,
+    {
+        Filter::enabled(&self.target.filter, meta, ctx)
+    }
+}
+
+impl<'a> MakeWriter for &'a LogTargetMakeWriter {
+    type Writer = LogOutput<'a>;
+
+    fn make_writer(&self) -> Self::Writer {
+        match &self.target.dest {
+            LogDest::Stdout => LogOutput::Stdout(self.stdout.lock()),
+            LogDest::File(file) => LogOutput::File(file.lock()),
+        }
+    }
+}
+
+/// A [`tracing_subscriber::Layer`] for replicating the logging functionality in
+/// [`bee_common`](bee_common::logger) without using the [`log`] crate as the global subscriber.
+///
+/// Without this layer, enabling this crate's [`Subscriber`] will disable all logging of any kind, since
+/// it will be used as the global subscriber for the lifetime of the program, and all [`log`] events will
+/// be ignored.
+///
+/// This layer registers an interest in [`Event`](tracing::Event)s that describe [`log`] events,
+/// generated by [`tracing_log`]. These are only created when
+/// [`collect_logs`](crate::subscriber::collect_logs) is called, or a [`LogTracer`](tracing_log::LogTracer)
+/// is initialised.
+pub struct LogLayer {
+    make_writers: Vec<LogTargetMakeWriter>,
+    fmt_events: LogFormatter,
+}
+
+impl<S> Layer<S> for LogLayer
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
```

Comment:
Doing fancy HKT bounds and all :trollface:

Reply:
😩